src/actions/plugins/imswitcher.ts

```diff
@@ -0,0 +1,138 @@
+    if (configuration.autoSwitchInputMethod.enable !== true) {
+      return;
+    }
+    if (!this.isConfigurationValid()) {
+      this.disableIMSwitch();
+      return;
+    }
+    // when you exit from insert-like mode, save origin input method and set it to default
+    let isPrevModeInsertLike = this.isInsertLikeMode(prevMode);
+    let isNewModeInsertLike = this.isInsertLikeMode(newMode);
+    if (isPrevModeInsertLike !== isNewModeInsertLike) {
+      if (isNewModeInsertLike) {
+        this.resumeIM();
+      } else {
+        this.switchToDefaultIM();
+      }
+    }
+  }
+
+  // save origin input method and set input method to default
+  private async switchToDefaultIM() {
+    const obtainIMCmd = configuration.autoSwitchInputMethod.obtainIMCmd;
+    const rawObtainIMCmd = this.getRawCmd(obtainIMCmd);
+    if (existsSync(rawObtainIMCmd) || Globals.isTesting) {
+      try {
+        const insertIMKey = await this.execute(obtainIMCmd);
+        if (insertIMKey !== undefined) {
+          this.savedIMKey = insertIMKey.trim();
+        }
+      } catch (e) {
+        logger.error(`IMSwitcher: promise is rejected. err=${e}`);
+      }
+    } else {
+      this.showCmdNotFoundErrorMessage(rawObtainIMCmd, 'vim.autoSwitchInputMethod.obtainIMCmd');
+    }
+
+    const defaultIMKey = configuration.autoSwitchInputMethod.defaultIM;
+    if (defaultIMKey !== this.savedIMKey) {
+      this.switchToIM(defaultIMKey);
+    }
+  }
+
+  // resume origin inputmethod
+  private async resumeIM() {
+    if (this.savedIMKey !== configuration.autoSwitchInputMethod.defaultIM) {
+      this.switchToIM(this.savedIMKey);
+    }
+  }
+
+  private async switchToIM(imKey: string) {
+    let switchIMCmd = configuration.autoSwitchInputMethod.switchIMCmd;
+    const rawSwitchIMCmd = this.getRawCmd(switchIMCmd);
+    if (existsSync(rawSwitchIMCmd) || Globals.isTesting) {
+      if (imKey !== '' && imKey !== undefined) {
+        switchIMCmd = switchIMCmd.replace('{im}', imKey);
+        try {
+          await this.execute(switchIMCmd);
+        } catch (e) {
+          logger.error(`IMSwitcher: promise is rejected. err=${e}`);
+        }
+      }
+    } else {
+      this.showCmdNotFoundErrorMessage(rawSwitchIMCmd, 'vim.autoSwitchInputMethod.switchIMCmd');
+    }
+  }
+
+  private isInsertLikeMode(mode: ModeName): boolean {
+    const insertLikeModes = new Set([
+      ModeName.Insert,
+      ModeName.Replace,
+      ModeName.SurroundInputMode,
+    ]);
+    if (insertLikeModes.has(mode)) {
+      return true;
+    }
+    return false;
+  }
+
+  private getRawCmd(cmd: string): string {
+    return cmd.split(' ')[0];
+  }
+
+  private showCmdNotFoundErrorMessage(cmd: string, config: string) {
+    Message.ShowError(`Unable to find ${cmd}. check your ${config} in VSCode setting.`);
+    this.disableIMSwitch();
+  }
+
+  private disableIMSwitch() {
+    const originConfig = configuration.autoSwitchInputMethod;
+    configuration.autoSwitchInputMethod = {
```

Comment:
Since these are separate configurations now, I think we can just go...

configuration.autoSwitchInputMethod.enable = false

Reply:
Can you add a test case for testing the scenarios with bad configs?

Reply:
Can you step through the code and see what operation is failing? We already have this pattern for a couple of other configurations like https://github.com/VSCodeVim/Vim/blob/master/src/configuration/configuration.ts#L247 so not sure why it wouldn't work.

Reply:
ok, I'll try to figure it out tonight.

Reply:
I found out the reason, the `vscode.workspace.getConfiguration()` function return a Proxy object instead of a object. The Proxy object's set method is replaced. So I got the `Cannot assign to read only property 'enable' of object.` error.

I've add commits to unwrap the Proxy object.